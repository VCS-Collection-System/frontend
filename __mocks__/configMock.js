export const getConfig = (URL, configUrl, config) => {
  if(URL.indexOf(configUrl) > -1 ) {
    const country = URL.substring(configUrl.length);
    const countryConfig = config[country] || null;
    return Promise.resolve({data: countryConfig});
  } else {
    return Promise.reject({data: null});
  }
}