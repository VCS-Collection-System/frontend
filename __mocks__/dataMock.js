export const employeeInfo = {
  "employeeId":"mock1",
  "firstName":"Mock",
  "lastName":"Tester",
  "email":"test@mock.com",
  "agencyCode":"Department of Mock Safety",
  "agencyName":"Department of Mock Safety",
  "divisionCode":"App Mocking",
  "divisionName":null,
  "supervisor":null,
  "dateOfBirth":"2021-10-07",
  "alternateEmail":"mock@email.com",
  "age":0,
  "ageInMonth":0,
  "workforceId":"mock2",
  "hr":false
}

export const attachmentObject = {
  "originalFileName":"blob",
  "size":100,
  "contentType":"image/png",
  "s3BucketName":"mockBucket",
  "s3UUID":"mockUUID"
}

export const vaxHistory = {
  "id":0,
  "employee": employeeInfo,
  "submissionDate":"2021-10-04T15:12:14.875",
  "review":{
     "reviewerEmployeeId":"z2",
     "reviewerNotes":"",
     "reviewDate":"2021-10-05T14:04:29.296",
     "reviewOutcome":"ACCEPTED",
     "rejectReason":""
  },
  "submittedBy":null,
  "autoApproved":false,
  "lastUpdatedTime":null,
  "lastUpdatedBy":null,
  "vaccineBrand":"MODERNA",
  "vaccineAdministrationDate":"2021-10-04",
  "fullVaccinatedFlag":false,
  "vaccineShotNumber":2,
  "documentType":"VaccineDocument",
  "attachment": {}
}

const taskEmployeeObject = {
  "employeeId":"mock1",
  "firstName":"Mock",
  "lastName":"Tester",
  "email":"test@mock.com",
  "agencyCode":"Department of Mock Safety",
  "agencyName":"Department of Mock Safety",
  "divisionCode":"App Mocking",
  "divisionName":null,
  "supervisor":null,
  "dateOfBirth": {
    "year": "2021",
    "monthValue": "12",
    "dayOfMonth": "16"
  },
  "alternateEmail":"mock@email.com",
  "age":0,
  "ageInMonth":0,
  "workforceId":"mock2",
  "hr":false
}

export const mockTask = {
  "task-input-data": {
    "documentList": {
      "com.redhat.vcs.model.VaccineDocumentList": {
        "documents": [ {
          "com.redhat.vcs.model.VaccineCardDocument": {
            "employee": {
              "com.redhat.vcs.model.Employee": taskEmployeeObject
            },
            "vaccineBrand": {
              "com.redhat.vcs.model.VaccineBrand": "MODERNA"
            },
            "attachment": {
              "com.redhat.vcs.model.Attachment": attachmentObject
            },
            "vaccineAdministrationDate": {
              "year": "2021",
              "monthValue": "12",
              "dayOfMonth": "16"
            },
            "lotNumber": "mockLot"
          } }
        }
      }
    }
  }
}
