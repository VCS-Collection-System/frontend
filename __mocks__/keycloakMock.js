module.exports = {
    token: "bogustoken", 
    authenticated: false,
    loadUserProfile: jest.fn( ()=> {return Promise.resolve({
        attributes: {
            workforceid: ["mockId"]
        }
    })} ),
    login: jest.fn(),
    logout: jest.fn(),
    hasRealmRole: jest.fn(),
    profile: {
        firstName: "mockFirst",
        lastName: "mockLast",
        email: "mock@email.com",
        attributes: {
            workforceid: ["mockId"],
            agency: ["USA"]
        }
    }
};
