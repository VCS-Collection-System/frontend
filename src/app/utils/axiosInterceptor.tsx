import axios from "axios";

const axiosInterceptor = axios.create({})

axiosInterceptor.interceptors.request.use(
    config => {
        config.headers['Authorization'] = 'Bearer ' + sessionStorage.getItem("token");
        return config;
    },
    error => {
        Promise.reject(error)
    });
export default axiosInterceptor;