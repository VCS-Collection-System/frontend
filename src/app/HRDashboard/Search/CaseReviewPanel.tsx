import React, { useEffect, useState } from 'react';
import { 
  Button,
  Card, 
  CardBody, 
  CardFooter,
  Grid,
  Flex,
  FlexItem,
  GridItem,
  Spinner,
  Split
} from '@patternfly/react-core';
import { CheckCircleIcon, ErrorCircleOIcon, ExternalLinkAltIcon } from '@patternfly/react-icons';
import CardWithTitle from '@app/Components/CardWithTitle';
import authaxios from "@app/utils/axiosInterceptor";
import ReviewInfoCardHorizontal from '@app/Components/ReviewInfoCardHorizontal';
import { ResponseErrorModal, GenericResponseModal } from '@app/Components/FeedbackModals';
import DeclineReasonModal from '@app/Components/DeclineReasonModal';
import { useKeycloak } from "@react-keycloak/web";
import { PAM_TASK_URL, VAX_TASK_DOCUMENT_URL, ATTACHMENT_URL, DELETE_VAX_URL } from '@app/utils/constants'
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';
import DeleteConfirmationModal from '@app/Components/DeleteConfirmationModal';


interface AppProps {
  hasSelectedItem: boolean;
  setHasSelectedItem: (boolean) => void; 
  updatePanel: boolean;
  refreshCaseHistory: boolean; 
  setResfreshCaseHistory: (boolean) => void; 
}

const CaseReviewPanel: React.FunctionComponent<AppProps> = ({
  hasSelectedItem, 
  setHasSelectedItem,
  updatePanel, 
  refreshCaseHistory, 
  setResfreshCaseHistory
}) => {
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const reviewerID = employee.id;
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const [isContentLoading, setIsContentLoading] = useState(false); 
  const [isDeclineModalOpen, setIsDeclineModalOpen] = useState(false);
  const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false); 
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false); 
  const [isGenericModalOpen, setIsGenericModalOpen] = useState(false);  
  const [errorState, setErrorState] = useState(''); 
  const [genericModalTitle, setGenericModalTitle] = useState("");
  const [genericModalMessage, setGenericModalMessage] = useState("");
  const [deleteModalTitle, setDeleteModalTitle] = useState("");
  const [deleteModalMessage, setDeleteModalMessage] = useState("");
  const [disableReview, setDisableReview] = useState(false); 

  const [submissionDate, setSubmissionDate] = useState(""); 
  const [historyType, setHistoryType] = useState("");
  const [testResult, setTestResult] = useState(""); 
  const [testDate, setTestDate] = useState("");
  const [reviewDate, setReviewDate] = useState(""); 
  const [vaxBrand, setVaxBrand] = useState("");
  const [vaxDate, setVaxDate] = useState(""); 
  const [vaxLotNumber, setVaxLotNumber] = useState("");
  const [vaxReview, setVaxReview] = useState("");
  const [reviewer, setReviewer] = useState(""); 
  const [taskId, setTaskId] = useState("");
  const [isPendingReview, setIsPendingReview] = useState(false)

  const [attachment, setAttachment] = useState(""); 
  const [attachmentType, setAttachmentType] = useState(""); 
  const [hasAttachment, setHasAttachment] = useState(false); 

  const [submittedBy, setSubmittedBy] = useState("");

  useEffect(() => {
    if(hasSelectedItem) {
      setIsContentLoading(true); 
      setIsPendingReview(false);
      setReviewer("");
      setDisableReview(false); 

      const reviewType = sessionStorage.getItem("caseHistoryType") || "";
      const sessionReivewData = sessionStorage.getItem("caseHistoryInfo") || "{}";
      const reviewData = JSON.parse(sessionReivewData);
      const attachmentProperty = reviewData['attachment'] || {};
      setHasAttachment(attachmentProperty['size'] !== 0)

      setSubmittedBy(reviewData['submittedBy']); 

      if(reviewType.includes('TEST')){
        setHistoryType("TEST")
        setTestResult(reviewData['covidTestResult']);
        setTestDate(reviewData['covidTestDate']);
        setSubmissionDate(returnNewDate(reviewData['submissionDate']));
      }
      else if(reviewType.includes('VAX')) {
        setHistoryType("VAX");
        setVaxBrand(reviewData['vaccineBrand']);
        setVaxDate(reviewData['vaccineAdministrationDate']);
        setVaxLotNumber(reviewData['lotNumber'])
        setSubmissionDate(returnNewDate(reviewData['submissionDate']));

        if(reviewData['review']){
          setVaxReview(reviewData['review']['reviewOutcome'])
          setReviewer(reviewData['review']['reviewerEmployeeId']); 
          setReviewDate(returnNewDate(reviewData['review']['reviewDate']));
        }
        else {
          if(!reviewData['autoApproved']) {
            setVaxReview("NOT REVIEWED");
            authaxios.get(VAX_TASK_DOCUMENT_URL + reviewData['id'])
            .then(res => {
              setIsPendingReview(true);
              setTaskId(res['data']['taskId']);
              authaxios.get(PAM_TASK_URL + res['data']['taskId'])
              .then(taskRes => {
                const taskData = taskRes['data']
                if(taskData['task-status'] === "Reserved") {
                  setDisableReview(true);
                  setGenericModalTitle("Task Was Claimed");
                  setGenericModalMessage("Please select another one.");
                  setIsGenericModalOpen(true);
                }
                else {
                  setDisableReview(false); 
                }
              })
            })
          }
          else{
            setVaxReview("AUTO-ACCEPTED");
          }
        }        
      }

      if(reviewType.includes('TEST') || reviewType.includes('VAX')) {
        authaxios.post(ATTACHMENT_URL, attachmentProperty, {
          responseType: 'arraybuffer'
        })
        .then(res => {
          setAttachment(res['data']);
          setAttachmentType(res.headers['content-type']);
        })
        .finally(() => {
          setIsContentLoading(false); 
        })
      }
      else {
        setIsContentLoading(false);
      }
    }
  }, [hasSelectedItem, updatePanel]);

  const handleDeleteRecord = () => {
    setIsFormSubmitting(true);
   
    const sessionReivewData = sessionStorage.getItem("caseHistoryInfo") || "{}";
    const recordId = JSON.parse(sessionReivewData)["id"];

    const headers = {
      'Accept': 'application/json'
    }

    authaxios.delete(DELETE_VAX_URL + recordId, {
      headers: headers
    }).then(() => {
        setHasSelectedItem(false);
        setDeleteModalTitle("Deleted Successfully!");
        setDeleteModalMessage("Record has been Deleted!");
    }).catch((res) => {
      if (res.response) {
        setIsFormSubmitting(false);
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    });

    setIsDeleteModalOpen(false);
  }

  const sendReview = (accepted, reason) => {
    setIsFormSubmitting(true)
    setIsDeclineModalOpen(false);
    const taskUrl = PAM_TASK_URL + taskId + "/states"
    const outcome = accepted ? "ACCEPTED" : "DECLINED"

    const headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    const acceptData = {
      "documentReview": {
        "com.redhat.vcs.model" : {
            "reviewerEmployeeId": reviewerID,
            "reviewOutcome": outcome,
            "reviewerNotes": reason,
            "rejectReason": reason
        }        
      }
    }

    authaxios.put(taskUrl + "/claimed", {}, {
      headers: headers
    })
    .then(() => {
      authaxios.put(taskUrl + "/completed?auto-progress=true", acceptData, {
        headers: headers
      })
      .then(() => {
        setHasSelectedItem(false);
        setGenericModalTitle("Submitted Successfully!");
        setGenericModalMessage("Review has been submitted!");
        setIsGenericModalOpen(true); 
      })
      .catch(submitResult => {
        if(submitResult.response) {
          setErrorState(submitResult.response.status);
          setIsErrorModalOpen(true);
        }
      })
    })
    .catch(claimResult => {
      if(claimResult.response) {
        setErrorState(claimResult.response.status);
        setIsErrorModalOpen(true);
      }
    }).finally(() => {
      setIsFormSubmitting(false)
    })
  }

  function returnNewDate(resDate) {
    const longDate = new Date(resDate);
    return ("0"+(longDate.getMonth()+1)).slice(-2) + "-" + ("0"+ longDate.getDate()).slice(-2) + "-" + longDate.getFullYear();
  }

  const closeReviewAndRefresh = () => {
    setResfreshCaseHistory(!refreshCaseHistory);
    setIsGenericModalOpen(false);
  }

  const ReviewInfo = () => {
    if (historyType === "TEST") {
      return (             
        <ReviewInfoCardHorizontal
          reviewType="TEST"
          covidTestResult={testResult}
          covidTestDate={testDate}
          submittedBy={submittedBy}
          submissionDate={submissionDate}
          reviewer={reviewer}
        />
      )
    }
    else if (historyType === "VAX") {
      return (
        <ReviewInfoCardHorizontal
          reviewType="VAX"
          vaccineBrand={vaxBrand}
          vaccineDate={vaxDate}
          vaccineLotNumber={vaxLotNumber}
          submittedBy={submittedBy}
          submissionDate={submissionDate}
          reviewer={reviewer}
          reviewDate={reviewDate}
          reviewStatus={vaxReview}
        />
      )
    }
    else {
      return null;
    }
  }

  const CaseAttachment = () => {
    if(attachmentType == "application/pdf") {
      return (
        <Button
          icon={<ExternalLinkAltIcon />}
          variant="link"
          component="a"
          href={URL.createObjectURL(new Blob([attachment], {type: 'application/pdf'}))}
          target="_blank"
        >
          Vaccination Card PDF
        </Button>
      )
    }
    else {
      return (<img src={URL.createObjectURL(new Blob([attachment]))} style={{maxHeight: '500px'}}/>)
    }
  }
  const DeleteContentButton = () => {
    if(hasSelectedItem) {
      return (
        <Button 
        variant="danger"
        onClick={() => setIsDeleteModalOpen(true)}
        style={{ marginLeft: "80%" }}
        >
          Delete Record</Button>
      )
    } else {
      return null;
    }
  }

  const CardContent = () => {
    if(!hasSelectedItem) {
      return (
        <CardBody>
          {"Please Select an Item"}
        </CardBody>
      )
    }
    else if(isContentLoading) {
      return (
        <CardBody isFilled>
          <Flex justifyContent={{ default: "justifyContentCenter" }}>
            <FlexItem>
              <Spinner isSVG/>
            </FlexItem>
          </Flex>
        </CardBody>
      )
    }
    else {
      return (
        <>
          <CardBody>
            <Grid hasGutter>
              <GridItem span={12}> 
                <ReviewInfo />
              </GridItem>
              <Flex justifyContent={{ default: "justifyContentCenter" }}>
                <FlexItem>
                  <GridItem span={12}> 
                    {
                      !hasAttachment ? "No Image" : <CaseAttachment />
                    }
                  </GridItem>
                </FlexItem>
              </Flex>
            </Grid>
          </CardBody>
          <CardFooter>
            {
              isPendingReview &&
              <Split hasGutter>
                <Button 
                  icon={<CheckCircleIcon />} 
                  isDisabled={disableReview}
                  variant="primary" 
                  onClick={() => sendReview(true, "")}
                >
                  Accept
                </Button>
                <Button 
                  icon={<ErrorCircleOIcon />} 
                  isDisabled={disableReview}
                  variant="danger" 
                  onClick = { () => setIsDeclineModalOpen(true) }
                >
                  Decline
                </Button>
              </Split>
            }
          </CardFooter>
        </>
      )
    }
  }

  return (
    <Card isRounded height="500px">
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <DeclineReasonModal 
        isModalOpen={isDeclineModalOpen}
        isSendingReview={isFormSubmitting}
        setIsModalOpen={setIsDeclineModalOpen}
        sendReview={sendReview}
      />
      <GenericResponseModal 
        isModalOpen={isGenericModalOpen}
        setIsModalOpen={closeReviewAndRefresh}
        title={genericModalTitle}
        message={genericModalMessage}
      />
      <DeleteConfirmationModal 
        isModalOpen={isDeleteModalOpen}
        setIsModalOpen={setIsDeleteModalOpen}
        title={deleteModalTitle}
        message={deleteModalMessage}
        handleDelete={handleDeleteRecord}
      />
      <CardWithTitle title="Attestation Details"/>
      <CardContent />
      <CardFooter>
        <Split hasGutter>
          <DeleteContentButton />
        </Split>
      </CardFooter>
    </Card>
  );
}

export { CaseReviewPanel };
