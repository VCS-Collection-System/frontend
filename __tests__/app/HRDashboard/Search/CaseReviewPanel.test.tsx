import React from 'react';
import { CaseReviewPanel } from '../../../../src/app/HRDashboard/Search/CaseReviewPanel';
import mockKeycloak from '../../../../__mocks__/keycloakMock'
import { act } from 'react-dom/test-utils';
import { ATTACHMENT_URL, VAX_TASK_DOCUMENT_URL, PAM_TASK_URL } from '../../../../src/app/utils/constants'
import { fireEvent, render, screen } from '@testing-library/react'
import reactRouterDom from 'react-router-dom'
import { employeeInfo, attachmentObject} from '../../../../__mocks__/dataMock'

jest.mock('react-router-dom')
const mockPush = jest.fn()
reactRouterDom.useHistory = jest.fn().mockReturnValue({push: mockPush})

jest.mock('@react-keycloak/web', () => ({
    useKeycloak: jest.fn( ()=>({keycloak: mockKeycloak}))
})) 

jest.mock("@app/utils/constants", () => ({
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  APPROVER_ROLES: ["mock-approver"],
  FEATURE_FLAGS: ["fakeFeature"],
  CONSENT_COUNTRIES: ["AUS"],
  RECORD_TYPES: {"USA": [{"value": "CDC", "label": "CDC Vaccination Card"}]},
  VAX_TASK_DOCUMENT_URL: "vax_task_url/",
  PAM_TASK_URL: "pam_task_url/"
}))

jest.mock('@app/utils/axiosInterceptor', () => ({
  get: (URL) => {
    let returnData; 
    if(URL === VAX_TASK_DOCUMENT_URL + "1") {
      returnData = {data: {taskId: "mockID"}}
    }
    else if(URL === VAX_TASK_DOCUMENT_URL + "2") {
      returnData = {data: {taskId: "mockIDTwo"}}
    }
    else if(URL === VAX_TASK_DOCUMENT_URL + "3") {
      returnData = {data: {taskId: "mockIDThree"}}
    }
    else if(URL === VAX_TASK_DOCUMENT_URL + "4") {
      returnData = {data: {taskId: "mockIDFour"}}
    }
    else if(URL === VAX_TASK_DOCUMENT_URL + "5") {
      returnData = {data: {taskId: "mockIDFive"}}
    }
    else if(URL === VAX_TASK_DOCUMENT_URL + "6") {
      returnData = {data: {taskId: "mockIDSix"}}
    }
    else if(URL === PAM_TASK_URL + "mockID"){
      returnData = {data: {"task-status": "ready"}}
    }
    else if(URL === PAM_TASK_URL + "mockIDTwo"){
      returnData = {data: {"task-status": "Reserved"}}
    }
    else if(URL === PAM_TASK_URL + "mockIDThree"){
      returnData = {data: {"task-status": "ready"}}
    }
    else if(URL === PAM_TASK_URL + "mockIDFour"){
      returnData = {data: {"task-status": "ready"}}
    }
    else if(URL === PAM_TASK_URL + "mockIDFive"){
      returnData = {data: {"task-status": "ready"}}
    }
    else if(URL === PAM_TASK_URL + "mockIDSix"){
      returnData = {data: {"task-status": "ready"}}
    }
    else{
      returnData = {data: []};
    }

    return Promise.resolve(returnData)
  },
  post: (URL, props) => {
    let returnData; 

    if(URL === ATTACHMENT_URL && props.originalFileName === 'mockPDF'){
      returnData = {data: {}, headers: {"content-type": "application/pdf"}};
    }
    else if(URL === ATTACHMENT_URL){
      returnData = {data: {}, headers: {}};
    }
    return Promise.resolve(returnData)
  },
  put: (URL) => {
    let returnData; 

    if(URL === PAM_TASK_URL + "mockID" + "/states/claimed"){
      returnData = {data: {}}
    }
    else if(URL === PAM_TASK_URL + "mockID" + "/states/completed?auto-progress=true"){
      returnData = {data: {}}
    }
    else if(URL === PAM_TASK_URL + "mockIDThree" + "/states/claimed"){
      return Promise.reject({response: {status: 500}})
    }
    else if(URL === PAM_TASK_URL + "mockIDFour" + "/states/claimed"){
      returnData = {data: {}}
    }
    else if(URL === PAM_TASK_URL + "mockIDFour" + "/states/completed?auto-progress=true"){
      return Promise.reject({response: {status: 500}})
    }
    else if(URL === PAM_TASK_URL + "mockIDFive" + "/states/claimed"){
      return Promise.reject({})
    }
    else if(URL === PAM_TASK_URL + "mockIDSix" + "/states/claimed"){
      returnData = {data: {}}
    }
    else if(URL === PAM_TASK_URL + "mockIDSix" + "/states/completed?auto-progress=true"){
      return Promise.reject({})
    }
    else{
      returnData = {data: {}};
    }
    return Promise.resolve(returnData)
  }
}))

const mockCaseHistoryInfo = {
  "id":1,
  "employee": employeeInfo,
  "attachment": attachmentObject,
  "submissionDate":"2021-11-08T15:44:47.06042",
  "review":null,
  "submittedBy":null,
  "autoApproved":true,
  "vaccineBrand":"MODERNA",
  "vaccineAdministrationDate":"2021-11-08",
  "vaccineShotNumber":1,
  "lotNumber":"1",
  "proofType":"CDC"
}

beforeEach(() => {
  sessionStorage.setItem("caseHistoryInfo", JSON.stringify(mockCaseHistoryInfo))
  sessionStorage.setItem("caseHistoryType", "VAX")
  window.URL.createObjectURL = jest.fn();
})

afterEach(() => {
  sessionStorage.clear(); 
})

const mockSetHasSelectedItem = jest.fn(); 
const mockSetRefreshCaseHistory = jest.fn(); 

describe('<CaseReviewPanel />', ()=> {
  it('should be defined', () => {
    expect(CaseReviewPanel).toBeDefined()
  })

  it("should match existing snapshot", async () => {
    sessionStorage.setItem("caseHistoryType", "MOCK")
    await act(async () => {
      const { asFragment } = render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })

  it("should display Vax: Auto-Accepted", async () => {
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    expect(screen.getByDisplayValue('AUTO-ACCEPTED')).toBeTruthy()
  })

  it("should display Vax: NOT REVIEWED", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    expect(screen.getByDisplayValue('NOT REVIEWED')).toBeTruthy()
  })

  it("should display Vax: NOT REVIEWED and reserved", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=2
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    expect(screen.getByText('Task Was Claimed')).toBeTruthy();

    const closeButton = screen.getAllByRole('button', {name: /Close/i})[0]

    await act(async () => {
      fireEvent.click(closeButton)
    })

    expect(screen.queryAllByText('Task Was Claimed')).toEqual([]);
  })

  it("should be able to accept a task", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=1
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    const acceptButton = screen.getByRole('button', {name: /Accept/i})

    await act(async () => {
      fireEvent.click(acceptButton)
    })

    expect(screen.getByText('Submitted Successfully!')).toBeTruthy();
  })

  it("should be able to decline a task", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=1
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    const declineButton = screen.getByRole('button', {name: /Decline/i})

    await act(async () => {
      fireEvent.click(declineButton)
    })

    expect(screen.getByText('Attachments are not legible')).toBeTruthy();

    await act(async () => {
      fireEvent.click(screen.getByRole('radio', {name: /Attachments are not legible/i}))
      fireEvent.click(screen.getByRole('button', {name: /Confirm/i}))
    })

    expect(screen.getByText('Submitted Successfully!')).toBeTruthy();
  })

  it("should display claim errors", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=3
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    const acceptButton = screen.getByRole('button', {name: /Accept/i})

    await act(async () => {
      fireEvent.click(acceptButton)
    })

    expect(screen.getByText('Server is unable to respond. Please try again later.')).toBeTruthy();
  })

  it("should not display claim errors if error does not return an error", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=5
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    const acceptButton = screen.getByRole('button', {name: /Accept/i})

    await act(async () => {
      fireEvent.click(acceptButton)
    })

    expect(screen.queryAllByText('Server is unable to respond. Please try again later.')).toEqual([])
  })

  it("should display submit errors", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=4
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    const acceptButton = screen.getByRole('button', {name: /Accept/i})

    await act(async () => {
      fireEvent.click(acceptButton)
    })

    expect(screen.getByText('Server is unable to respond. Please try again later.')).toBeTruthy();
  })

  it("should not display submit errors if server does not return an error", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.autoApproved=false;
    expectData.id=6
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    const acceptButton = screen.getByRole('button', {name: /Accept/i})

    await act(async () => {
      fireEvent.click(acceptButton)
    })

    expect(screen.queryAllByText('Server is unable to respond. Please try again later.')).toEqual([])
  })

  it("should display pdf", async () => {
    const expectData = mockCaseHistoryInfo;
    expectData.attachment = {
      "originalFileName":"mockPDF",
      "size":100,
      "contentType":"image/png",
      "s3BucketName":"mockBucket",
      "s3UUID":"mockUUID"
    }
    sessionStorage.setItem("caseHistoryInfo", JSON.stringify(expectData))
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    expect(screen.getByText('Vaccination Card PDF')).toBeTruthy()
  })

  it("should not display anything if no data in session", async () => {
    sessionStorage.clear()
    await act(async () => {
      render(
        <CaseReviewPanel 
          hasSelectedItem={true}
          setHasSelectedItem={mockSetHasSelectedItem}
          updatePanel={false}
          refreshCaseHistory={false}
          setResfreshCaseHistory={mockSetRefreshCaseHistory}
        />
      )
    })

    expect(screen.queryAllByText('NOT REVIEWED')).toEqual([]);
    expect(screen.queryAllByText('ACCEPTED')).toEqual([]);
    expect(screen.queryAllByText('DECLINED')).toEqual([]);
    expect(screen.queryAllByText('AUTO-ACCEPTED')).toEqual([]);
  })
})