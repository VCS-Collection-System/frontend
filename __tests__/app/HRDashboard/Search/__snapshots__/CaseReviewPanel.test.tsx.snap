// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CaseReviewPanel /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card pf-m-rounded"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    height="500px"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      <p
        class=""
        data-pf-content="true"
      >
        Attestation Details
      </p>
      <div
        class="pf-c-content"
      >
        <small
          class=""
          data-pf-content="true"
        />
      </div>
    </div>
    <div
      class="pf-c-card__body"
    >
      <div
        class="pf-l-grid pf-m-gutter"
      >
        <div
          class="pf-l-grid__item pf-m-12-col"
        />
        <div
          class="pf-l-flex pf-m-justify-content-center"
        >
          <div
            class=""
          >
            <div
              class="pf-l-grid__item pf-m-12-col"
            >
              No Image
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-c-card__footer"
    />
    <div
      class="pf-c-card__footer"
    >
      <div
        class="pf-l-split pf-m-gutter"
      >
        <button
          aria-disabled="false"
          class="pf-c-button pf-m-danger"
          data-ouia-component-id="OUIA-Generated-Button-danger-1"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe="true"
          style="margin-left: 80%;"
          type="button"
        >
          Delete Record
        </button>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
